File: baekjoon/silver2/5397.py
# input n (execute times)
n = int(input())

# execute n times
for _ in range(n):
  leftside = [] # left of cursor
  rightside = [] # right of cursor

  password = input()
  for c in password:
    # backspace : pop from left stack
    if c == '-': 
      # check cursor is at the beginning
      if(leftside):
        leftside.pop()

    # left arrow : move char left stack to right stack
    elif c == '<': 
      if(leftside): # check leftside is empty
        rightside.append(leftside.pop())

    # right arrow : move char right stack to left stack
    elif c == '>':
      if(rightside): # check rightside is empty
        leftside.append(rightside.pop())

    # number or letter : add left stack 
    else:
      leftside.append(c)

  leftside.extend(reversed(rightside))
  print(''.join(leftside))

File: baekjoon/silver2/5397.cc
#include <iostream>
#include <list>
#include <string>
using namespace std;

// list of password key
list<char> password;

// get correct password
void get_correct_key(string pwd) {
  list<char>::iterator cursor = password.begin();
  
  for(char c : pwd) {
    switch(c){
      // backspace
      case '-':
        // check cursor is at the beginning
        if(cursor != password.begin()) {
          cursor = password.erase(--cursor);
        }
        break;

      // left arrow
      case '<':
        // check cursor is at the beginning
        if(cursor != password.begin()) {
          cursor--;
        }
        break;

      // right arrow
      case '>':
        // check cursor is at the end
        if(cursor != password.end()) {
          cursor++;
        }
        break;

      // numbers and letters
      default:
        password.insert(cursor, c);
    }
  }
}

// print password (list -> string)
void print_password() {
  for(char key : password) {
    cout << key;
  }
  cout << endl;
}

int main() {
  // input # of testcase
  int testcase = 0;
  cin >> testcase;

  // store input password
  string input_pwd = "";

  // execute #testcase times
  for(int i = 0; i < testcase; i++) {
    // input password
    cin >> input_pwd;

    // initialize password list
    password.clear();

    // get password
    get_correct_key(input_pwd);
  
    // print result
    print_password();
  }
  
  return 0;
}

